### PR TITLE
[TP] Enable Dropout for DT and add test for activations

### PR DESCRIPTION
**Pull request by fduwjj**
Enable dropout for DT.

### CODE REVIEW

File: test/spmd/tensor/test_ops.py [hunk 1]
```diff
@@ -102,6 +103,43 @@ def test_ones_like(self):
...
+        self._run_sharded_elementwise_ops(device_mesh, [Shard(0)], (8, 5),
+                                          torch.nn.functional.dropout, p=0.4, training=False)
+        self._run_sharded_elementwise_ops(device_mesh, [Shard(1)], (3, 14),
+                                          torch.nn.functional.dropout, p=0.5, training=True)
+
+    @with_comms
+    def test_dropout_errors(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        with self.assertRaisesRegex(RuntimeError, 'Not supported!'):
+            self._run_sharded_elementwise_ops(device_mesh, [_Partial(ReduceOp.SUM)], (8, 5),
```
Comment (fduwjj) [inline test/spmd/tensor/test_ops.py:140]: @wanchaol Looks like partial tensor does not trigger the override. So no errors are thrown here.

File: spmd/tensor/ops/dropout.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+from typing import Tuple
+from spmd.tensor.ops.utils import register_impl
+from spmd.tensor.api import Tensor
+
+
+@register_impl("aten.native_dropout.default")
+def _dist_dropout(self: Tensor, p: float, train: bool) -> Tuple[Tensor, Tensor]:
+    print(self.placements)
+    self_placement = self.placements[0]
+    if self_placement.is_partial():
```
Comment (wanchaol) [inline spmd/tensor/ops/dropout.py:12]: let's error out on both partial and replicated for now, and maybe add an TODO here to explain the CPU/GPU difference and fix later

File: test/spmd/tensor/test_ops.py [hunk 3]
```diff
@@ -102,6 +103,36 @@ def test_ones_like(self):
...
+    @with_comms
+    def test_activations(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        self._run_sharded_elementwise_ops(device_mesh, [Shard(0)], (8, 5), torch.nn.functional.gelu)
+        self._run_sharded_elementwise_ops(device_mesh, [Replicate()], (8, 5), torch.nn.functional.gelu)
+        self._run_sharded_elementwise_ops(device_mesh, [Shard(1)], (3, 14), torch.nn.functional.relu)
+        self._run_sharded_elementwise_ops(device_mesh, [Replicate()], (8, 5), torch.nn.functional.relu)
+
+    @with_comms
+    def test_dropout(self):
```
Comment (wanchaol) [inline test/spmd/tensor/test_ops.py:129]: can we only test this for GPU? also add a TODO here to enable CPU case so we don't forget haha

File: spmd/tensor/ops/dropout.py (hunk 2 above)
Comment (fduwjj) [reply spmd/tensor/ops/dropout.py:12]: We can still keep this logic since we need this logic for GPU right?

File: test/spmd/tensor/test_ops.py (hunk 3 above)
Comment (fduwjj) [reply test/spmd/tensor/test_ops.py:129]: Sure.

File: spmd/tensor/ops/dropout.py (hunk 2 above)
Comment (wanchaol) [reply spmd/tensor/ops/dropout.py:12]: Yep, in GPU it will trigger this function

File: spmd/tensor/ops/dropout.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from spmd.tensor.ops.utils import register_impl
+from spmd.tensor.api import Tensor
+
+
+@register_impl("aten.native_dropout.default")
+def _dist_dropout(self: Tensor, p: float, train: bool) -> Tuple[Tensor, Tensor]:
+    self_placement = self.placements[0]
+    # TODO: To figure out why partial tensor does not dispatch here when in CPU.
+    # and with kwargs.
+    if self_placement.is_partial():
```
Comment (wanchaol) [inline spmd/tensor/ops/dropout.py:13]: I think we should error out now for the replicated case, as it's non-deterministic for replicated output. We can enforce the seed generation to be the same for replicated tensor in later PRs

File: test/spmd/tensor/test_elementwise_ops.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        reset_seed() if reset_seed else None
+        expected = op(input_tensor, **kwargs)
+        self.assertEqual(input_tensor, dist_tensor.local_tensor())
+        self.assertEqual(expected, dt.local_tensor())
+
+    @with_comms
+    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    def test_activations(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        self._run_sharded_elementwise_ops_gpu(device_mesh, [Shard(0)], (8, 5), torch.nn.functional.gelu)
```
Comment (wanchaol) [inline test/spmd/tensor/test_elementwise_ops.py:31]: did we figure out why gelu/relu works magically

**Review by wanchaol — APPROVED** [review]
lgtm, two comments.

**Comment by wanchaol** [comment]
nit: linting seems failing on CI

File: test/spmd/tensor/test_elementwise_ops.py (hunk 5 above)
Comment (fduwjj) [reply test/spmd/tensor/test_elementwise_ops.py:31]: No idea tbh..

File: spmd/tensor/ops/dropout.py (hunk 4 above)
Comment (fduwjj) [reply spmd/tensor/ops/dropout.py:13]: Got it, sure.
